src/coreclr/jit/rangecheckcloning.cpp

```diff
@@ -388,6 +388,56 @@ class BoundsChecksVisitor final : public GenTreeVisitor<BoundsChecksVisitor>
     }
 };
 
+// -----------------------------------------------------------------------------
+// DoesComplexityExceed: Check if the complexity of the bounds checks exceeds the budget.
+//    We want to avoid cloning blocks with too many unrelated trees/statements between
+//    the bounds checks.
+//
+// Arguments:
+//    bndChks - The stack of bounds checks
+//
+// Return Value:
+//    true if the complexity exceeds the budget, false otherwise.
+//
+static bool DoesComplexityExceed(ArrayStack<BoundsCheckInfo>* bndChks)
+{
+    Statement* firstBndChkStmt = bndChks->Bottom().stmt;
+    Statement* lastBndChkStmt  = bndChks->Top().stmt;
+
+    JITDUMP("Checking complexity from " FMT_STMT " to " FMT_STMT "\n", firstBndChkStmt->GetID(),
+            lastBndChkStmt->GetID());
+
+    // Some arbitrary limit just in case
+    if (bndChks->Height() > 64)
+    {
+        JITDUMP("\tExceeded the maximum number of bounds checks: %d\n", bndChks->Height());
+        return true;
+    }
+
+    // We're relying on GetCostSz to avoid walking the trees again.
+    // We assume the costs are initialized and relatively accurate.
```

Comment:
It seems ok to me to use `GetCostSz`, but @AndyAyersMS may have some feedback since he ended up not using it for loop cloning (why?). Also I wouldn't expect it to be that expensive to use `gtComplexityExceeds` since this code runs only  in the case where we are actually optimizing.

Reply:
Yeah it was not about saving TP (this PR has a minor impact on it), but the fact that CostSz takes into account the complexity of trees, not just their amount? Of course, if it's accurate/maintained